File: tools/datarunner/datarunner/ocr.py
"""CPU Tesseract plus optional NVIDIA/AMD/Intel GPU OCR (RapidOCR + ONNX)."""

from __future__ import annotations

import os
import shutil
import subprocess
from dataclasses import dataclass
from pathlib import Path


@dataclass(frozen=True)
class OcrResult:
    text: str
    backend: str
    device: str


def detect_gpu() -> str:
    """Return cuda | rocm | openvino | cpu from the host, not from pip extras."""
    if shutil.which("nvidia-smi") or Path("/dev/nvidia0").exists():
        return "cuda"
    if Path("/dev/kfd").exists() or shutil.which("rocminfo"):
        return "rocm"
    drm = Path("/sys/class/drm")
    if drm.is_dir():
        for child in drm.iterdir():
            uevent = child / "device" / "uevent"
            try:
                raw = uevent.read_text(encoding="utf-8", errors="ignore").lower()
            except OSError:
                continue
            if "i915" in raw or "xe\n" in raw or "driver=xe" in raw or "pci_id=8086" in raw:
                return "openvino"
    return "cpu"


def _tesseract(image: Path) -> str:
    exe = shutil.which("tesseract")
    if not exe:
        raise FileNotFoundError(
            "tesseract not found. Install the distro package, or pip install a GPU extra "
            "(ocr-cuda / ocr-rocm / ocr-intel) for RapidOCR."
        )
    proc = subprocess.run(
        [exe, str(image), "stdout", "--psm", "6", "-l", "eng"],
        check=False,
        capture_output=True,
        text=True,
    )
    if proc.returncode != 0:
        err = (proc.stderr or proc.stdout or "").strip()
        raise RuntimeError(f"tesseract failed: {err or proc.returncode}")
    return proc.stdout or ""


def _rapidocr(image: Path, device: str) -> str:
    """RapidOCR ONNX. Extra packages select the Execution Provider."""
    try:
        from rapidocr import RapidOCR  # type: ignore
    except ImportError as exc:
        raise ImportError(
            "RapidOCR is not installed. pip install -e '.[ocr-cuda]' (NVIDIA), "
            "'.[ocr-rocm]' (AMD), or '.[ocr-intel]' (Intel OpenVINO)."
        ) from exc

    params: dict[str, object] = {}
    if device == "cuda":
        params["EngineConfig.onnxruntime.use_cuda"] = True
    elif device == "rocm":
        # RapidOCR talks to whatever onnxruntime wheel is installed.
        os.environ.setdefault("ORT_ROCM_DEVICE_ID", "0")
    elif device in ("openvino", "intel"):
        params["EngineConfig.onnxruntime.use_openvino"] = True

    try:
        engine = RapidOCR(params=params) if params else RapidOCR()
        out = engine(str(image))
    except TypeError:
        engine = RapidOCR()
        out = engine(str(image))

    # RapidOCR 2 returns (result, elapse); 3 returns a result object.
    if isinstance(out, tuple):
        result = out[0]
    else:
        result = getattr(out, "txts", None) or getattr(out, "result", out)
    if result is None:
        return ""
    lines: list[str] = []
    if isinstance(result, list):
        for item in result:
            if isinstance(item, str):
                lines.append(item)
            elif isinstance(item, (list, tuple)) and len(item) >= 2:
                # [[box], text, score]
                text = item[1]
                if isinstance(text, str):
                    lines.append(text)
            elif isinstance(item, dict) and "text" in item:
                lines.append(str(item["text"]))
    elif isinstance(result, str):
        lines.append(result)
    return "\n".join(lines)


def image_to_text(image: Path, device: str = "auto") -> OcrResult:
    image = Path(image)
    if not image.is_file():
        raise FileNotFoundError(image)
    wanted = detect_gpu() if device == "auto" else device
    if wanted != "cpu":
        try:
            text = _rapidocr(image, wanted)
            return OcrResult(text=text, backend="rapidocr", device=wanted)
        except Exception as err:
            # GPU extra missing or EP failed — CPU Tesseract is the floor.
            fallback = _tesseract(image)
            return OcrResult(
                text=fallback + ("" if fallback.endswith("\n") else "\n")
                + f"[ocr: {wanted} failed ({err.__class__.__name__}); used tesseract]",
                backend="tesseract",
                device="cpu",
            )
    text = _tesseract(image)
    return OcrResult(text=text, backend="tesseract", device="cpu")

File: tools/datarunner/src/ocr.rs
//! CPU Tesseract; GPU RapidOCR if a `rapidocr` CLI is on PATH, else Tesseract.

use crate::config::OcrDevice;
use anyhow::{bail, Context, Result};
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;

#[derive(Clone, Debug)]
pub struct OcrResult {
    pub text: String,
    pub backend: &'static str,
    pub device: &'static str,
}

pub fn detect_gpu() -> OcrDevice {
    if which("nvidia-smi") || Path::new("/dev/nvidia0").exists() {
        return OcrDevice::Cuda;
    }
    if Path::new("/dev/kfd").exists() || which("rocminfo") {
        return OcrDevice::Rocm;
    }
    if intel_drm() {
        return OcrDevice::Openvino;
    }
    OcrDevice::Cpu
}

fn which(bin: &str) -> bool {
    env_path()
        .iter()
        .any(|dir| dir.join(bin).is_file() || dir.join(format!("{bin}.exe")).is_file())
}

fn env_path() -> Vec<PathBuf> {
    std::env::var_os("PATH")
        .map(|p| std::env::split_paths(&p).collect())
        .unwrap_or_default()
}

fn intel_drm() -> bool {
    let drm = Path::new("/sys/class/drm");
    let Ok(entries) = fs::read_dir(drm) else {
        return false;
    };
    for ent in entries.flatten() {
        let uevent = ent.path().join("device/uevent");
        let Ok(raw) = fs::read_to_string(uevent) else {
            continue;
        };
        let low = raw.to_ascii_lowercase();
        if low.contains("i915") || low.contains("driver=xe") || low.contains("pci_id=8086") {
            return true;
        }
    }
    false
}

fn tesseract(image: &Path) -> Result<String> {
    if !which("tesseract") {
        bail!(
            "tesseract not found. Install the distro package (tesseract / tesseract-ocr). \
             GPU RapidOCR is used only when a `rapidocr` binary is on PATH."
        );
    }
    let out = Command::new("tesseract")
        .arg(image)
        .arg("stdout")
        .arg("--psm")
        .arg("6")
        .arg("-l")
        .arg("eng")
        .output()
        .context("spawn tesseract")?;
    if !out.status.success() {
        let err = String::from_utf8_lossy(&out.stderr);
        bail!("tesseract failed: {}", err.trim());
    }
    Ok(String::from_utf8_lossy(&out.stdout).into_owned())
}

fn rapidocr_cli(image: &Path) -> Result<String> {
    if !which("rapidocr") {
        bail!("rapidocr CLI not on PATH");
    }
    let out = Command::new("rapidocr")
        .arg(image)
        .output()
        .context("spawn rapidocr")?;
    if !out.status.success() {
        bail!("rapidocr exit {}", out.status);
    }
    Ok(String::from_utf8_lossy(&out.stdout).into_owned())
}

pub fn image_to_text(image: &Path, device: OcrDevice) -> Result<OcrResult> {
    if !image.is_file() {
        bail!("{} is not a file", image.display());
    }
    let wanted = if device == OcrDevice::Auto {
        detect_gpu()
    } else {
        device
    };
    if wanted != OcrDevice::Cpu {
        match rapidocr_cli(image) {
            Ok(text) => {
                return Ok(OcrResult {
                    text,
                    backend: "rapidocr",
                    device: wanted.as_str(),
                });
            }
            Err(err) => {
                eprintln!(
                    "rapidocr failed on {} ({err}); using tesseract",
                    wanted.as_str()
                );
                let fallback = tesseract(image)?;
                return Ok(OcrResult {
                    text: fallback,
                    backend: "tesseract",
                    device: "cpu",
                });
            }
        }
    }
    Ok(OcrResult {
        text: tesseract(image)?,
        backend: "tesseract",
        device: "cpu",
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn detect_gpu_known_label() {
        let d = detect_gpu();
        assert!(matches!(
            d,
            OcrDevice::Cpu | OcrDevice::Cuda | OcrDevice::Rocm | OcrDevice::Openvino
        ));
    }
}
